src/db.c

```diff
@@ -1414,6 +1414,31 @@ int *zunionInterGetKeys(struct redisCommand *cmd, robj **argv, int argc, int *nu
     return keys;
 }
 
+/* Helper function to extract keys from following commands:
+ * ZUNION <num-keys> <key> <key> ... <key> <options>
+ * ZINTER <num-keys> <key> <key> ... <key> <options> */
+int *zunionInterGetKeys(struct redisCommand *cmd, robj **argv, int argc, int *numkeys) {
```

Comment:
We now have three functions that are basically, 
```<num-keys> <key> <key> ... ```
(EvalGetKeys/ZunionInter/ZunionInterStore)

Seems worth simplifying it now to a single helper function with 3 simpler functions for each type.

Reply:
I think the current code is not very redundant, although they are all `<num-keys> <key> <key> ...` pattern, but the details are different, for example:

- The `<num-keys>` arrry index of zunionInterStoreGetKeys is 2, but zunionInterGetKeys is 1, and evalGetKeys is 2.
- The number of keys is different, resulting in different malloc lengths.

If we merge the code, it may require multiple judgments for setting `num-keys`, for example:

```
    if (!strcasecmp(cmd->name, "ZINTER") || !strcasecmp(cmd->name, "ZUNION")) {
         keyIndex = 1
     } else if (!strcasecmp(cmd->name, "ZINTERSTORE") || !strcasecmp(cmd->name, "ZUNIONSTORE") ||
                !strcasecmp(cmd->name, "EVAL")) {
         keyIndex = 2;
     } 
```

This does not reduce much code, and there are many judgments, so my suggestion is to keep the current code.

Reply:
@yangbodong22011 i think what @madolson suggested is like so (no extra string matching):

```c
int *genericGetKeys(int storeKeyOfs, int keyCountOfs, int firstKeyOfs, int keyStep,
                    robj **argv, int argc, int *numkeys) {
...
...
}

int *zunionInterStoreGetKeys(struct redisCommand *cmd, robj **argv, int argc, int *numkeys) {
    return genericGetKeys(1, 2, 3, 1, argv, argc, numkeys);
}

int *zunionInterGetKeys(struct redisCommand *cmd, robj **argv, int argc, int *numkeys) {
    return genericGetKeys(-1, 1, 2, 1, argv, argc, numkeys);
}

int *evalGetKeys(struct redisCommand *cmd, robj **argv, int argc, int *numkeys) {
    return genericGetKeys(-1, 2, 3, 1, argv, argc, numkeys);
}
```

i guess this makes sense.. especially if we'll have more and more of these commands in the future.
maybe we should some day even move that info into the command table, and have it reflected in the COMMAND command (so that client libraries can more easily compute the hash slot without calling COMMAND GETKEYS).

Reply:
Oh, I see..., I agree with the optimization.

Reply:
already updated. Thanks @madolson  @oranagra